### DISCUSSION TITLE
Bootstrap, direct DOM manipulation & Vue's virtual DOM

### DISCUSSION THREAD

**Discussion by TechAkayy** [cdmoro/bootstrap-vue-3#310] [Q&A]
Hi there, 

Thank you for spearheading Bootstrap 5 + Vue 3 combo. 

I would like to ask a very fundamental question, surprisingly I haven't really seen this discussed anywhere online expect for this stackoverflow [comment](https://stackoverflow.com/questions/65547199/using-bootstrap-5-with-vue-3#comment125680515_66386443)

![image](https://user-images.githubusercontent.com/73772701/157210617-f40dae7e-d044-4769-923b-743e88fec64e.png)

I know mixing Vue (virtual dom) + Jquery/Vanilla JS (that does direct dom manipulator) is an [anti-pattern](https://forum.vuejs.org/t/how-vue-2-0-work-with-jquery/3591/3), [anti-pattern](https://forum.vuejs.org/t/when-is-it-ok-to-directly-manipulate-the-dom/45818/3). 

My question is does the bootstrap vue wrapper libraries like yours and the other bootstrap-vue one, do they all mitigate this anti-pattern by preventing direct dom manipulation by  bootstrap.min.js?

If so, would this be one of the reasons (on top of all the benefits & features) to use bootstrap-vue-3 against using bootstrap-5 directly in Vue 3 projects? 

Sorry, if it turns out to be a silly question, it's quite possible I might have misunderstood somethings! Thanks in advance.

**Comment by xvaara** [comment]
I kinda asked the same question in #243 

It took me about 30min to write the dropdown component as a "native" vue3 + popper.js component. It's just hard to move one component to native, since the way bootstrap.js is imported (at import it runs code to add eventlisteners to elements and so one).

Also don't know if all the compatibility with bootstrap.js way of using EventListeners and data- attributes is needed if written to use vue directly.

But it seems bootstrapVue is back from dead, so I don't know what the future of this project is. 
I have to run my own for of this project and bootstrap to be able to do SSR, and it's not fun. :(

**Reply by aceofwings** [reply]
Yeah interesting question. Some components are written without using their corresponding JS utilities. One reason being that the utilities often inherently add the anti pattern you mentioned above. Other times they stun the components ability in features and  ease of use . Ultimately I believe our direction in the long run will be to depend on more bootstraps styling.

**Comment by TechAkayy** [comment]
Thanks for sharing your thoughts. I have posted the same question on bootstrap-vue discussions - https://github.com/bootstrap-vue/bootstrap-vue/discussions/6931. It has around 13.7k stars & 1.8k forks. So, I'm hoping one them could clarify on our concerns, and bring closure on this topic. Cheers!

**Reply by mariusa** [reply]
Could you also ask at https://github.com/twbs/bootstrap/discussions

**Reply by TechAkayy** [reply]
Thxs for the suggestion @mariusa , I have posted it there as well - https://github.com/twbs/bootstrap/discussions/36011

**Comment by TechAkayy** [comment]
Also, I was wondering if frameworks without virtual-dom (like Svelte) also face the same issues. If not, then it's a big plus for bootstrap+svelte combo. 

Looks like bootstrap (js) combo with virtual-dom based frameworks is the real concern here. 

I wonder how much of bootstrap is used (I mean components with JS, not just CSS) in the other virtual-dom based React world!

**Comment by TechAkayy** [comment]
I might have found some clarification looking into react's equivalent of bootstrap which is https://react-bootstrap.github.io/, thxs to [this](https://github.com/twbs/bootstrap/discussions/36011) reply in the twbs bootstrap discussions. 

Here's an extract from react-bootstrap's [home page](https://react-bootstrap.github.io/). 
![image](https://user-images.githubusercontent.com/73772701/158054029-bd1a72a4-0519-4eb4-bfc1-ea335b0e9745.png)

So, basically, with react-bootstrap (quite an old one), it doesn't use bootstrap.js and replaces it with a react library of components (bootstrap styled). I wonder if bootstrap-vue & bootstrap-vue-3 also does something similar? 

Something probably out of scope for this repo - lately headless ui has been emerging.. 
- [headless ui - html/vue3/react](https://headlessui.dev/) (by tailwind creators but doesn't have anything tailwind related) + tailwindcss -> [tailwind-ui](https://tailwindui.com/) 
- [headless ui - html/vue3/react](https://headlessui.dev/) + bootstrap.css -> [bootstrap-ui](https://bootstrapui.com/) (could be)

Would be interesting to see an initiative building the *vue* part of bootstrap-ui (that doesn't exist yet) which could be a library of headless (vue) components that are bootstrap-styled!

Guessing this idea is in line with xvaara's idea here - https://github.com/cdmoro/bootstrap-vue-3/discussions/197

**Reply by xvaara** [reply]
I've been eyeing headless ui for awhile with the idea of doing bootstrap on top of it. The main problem is that we would loose a lot of functionality that we mimic from bootstrapVue. The components are a little bit more complex in the way they generate bootstrap classes. The actual functionality of the components isn't that complex. So using headless ui would not bring that much more to the table, and would require us to depend on their progress.
